include/deal.II/meshworker/scratch_data.h

```diff
@@ -0,0 +1,1301 @@
+     */
+    const std::vector<double> &
+    get_neighbor_JxW_values() const;
+
+    /**
+     * Return the last computed normal vectors.
+     */
+    const std::vector<Tensor<1, spacedim>> &
+    get_normal_vectors() const;
+
+    /**
+     * Return the last computed normal vectors on the neighbor.
+     */
+    const std::vector<Tensor<1, spacedim>> &
+    get_neighbor_normal_vectors();
+
+    /**
+     * Print the contents of the internal cache.
+     */
+    template <class Stream>
+    void
+    print_info(Stream &os);
+
+    /**
+     * Extract the local dof values associated with the internally initialized
+     * cell.
+     *
+     * Before you call this function, you have to make sure you have previously
+     * called one of the reinit() functions.
+     *
+     * At every call of this function, a new vector of dof values is generated
+     * and stored internally, unless a previous vector with the same name is
+     * found. If this is the case, the content of that vector is overwritten.
+     *
+     * If you give a unique @p global_vector_name, then for each cell you are
+     * guaranteed to get a unique vector of independent dofs of the same type
+     * as the dummy variable. If you use an automatic differentiation number
+     * type (like Sacado::Fad::DFad<double>,
+     * Sacado::Fad::DFad<Sacado::Fad::DFad<double>>, etc.) this method will
+     * also initialize the independent variables internally, allowing you
+     * to perform automatic differentiation.
+     *
+     * Once you have extracted the local dof values, you can later access the
+     * same dof values, without reinitializing them, by calling the method
+     * get_local_dof_values() with the same @p global_vector_name argument
+     * you passed here.
+     *
+     * Notice that using this initialization strategy renders the use of this
+     * ScratchData object incompatible with the AD helper classes (since they
+     * do their own data management). In particular, it is necessary for the
+     * user to manage all of the AD data themselves (both before and after this
+     * call).
+     */
+    template <typename VectorType, typename Number = double>
+    const std::vector<Number> &
+    extract_local_dof_values(const std::string &global_vector_name,
+                             const VectorType & input_vector,
+                             const Number       dummy = Number(0));
+
+    /**
+     * After calling extract_local_dof_values(), you can retrieve the stored
+     * information through this method.
+     *
+     * Both the argument @p global_vector_name and the type of the @p dummy
+     * variable should match the ones you passed to the
+     * extract_local_dof_values() function.
+     */
+    template <typename Number = double>
+    const std::vector<Number> &
+    get_local_dof_values(const std::string &global_vector_name,
+                         Number             dummy = Number(0)) const;
```

Comment:
same thing here about the dummy argument.

Reply:
and everywhere below, of course

Reply:
Same reasoning given above.